### ISSUE TITLE
Add explicit BroadcastOp for TTNN

### ISSUE THREAD

**Issue by uazizTT** [tenstorrent/tt-metal#16015]
Currently there is no support of broadcast for cases that requires it outside of binary or ternary eltwise operations. The requested BroadcastOp will perform equivalent to `torch.broadcast_to` Op.

Some front-ends require broadcast operation before unary ops. In this case having an explicit broadcastOp would be ideal. 

Following is an example from a front-end that would benefit from BroadcastOp. 

```
module {
  func.func @main(%arg0: tensor<6x2xf32>) -> tensor<2400x2xf32> {
    %0 = stablehlo.reshape %arg0 : (tensor<6x2xf32>) -> tensor<1x6x2xf32>
    %1 = stablehlo.reshape %0 : (tensor<1x6x2xf32>) -> tensor<1x6x1x2xf32>
    %2 = stablehlo.broadcast_in_dim %1, dims = [0, 1, 2, 3] : (tensor<1x6x1x2xf32>) -> tensor<400x6x1x2xf32>
    %3 = stablehlo.reshape %2 : (tensor<400x6x1x2xf32>) -> tensor<2400x1x2xf32>
    %4 = stablehlo.reshape %3 : (tensor<2400x1x2xf32>) -> tensor<2400x2xf32>
    return %4 : tensor<2400x2xf32>
  }
}
```

```
Torch equivalent to above is...

a = torch.arange(12).view(6, 2) 
print("Original tensor (a):")
print(a)

b = torch.broadcast_to(a, (2400 // 6, 6, 2))  # First, align with broadcasting
b = b.reshape(2400, 2)  # Reshape to desired final shape

print("\nBroadcasted tensor (b) shape:")
print(b.shape)
print("\nBroadcasted tensor (b):")
print(b)
```

**Comment by uazizTT** [comment]
> [@uazizTT](https://github.com/uazizTT) I suppose before the ticket is supported, the work around people use is ttnn.repeat()? For example, above operation can be done by ttnn.repeat(400, 1)?

@dchenTT Yes that's right, we use `ttnn.repeat`, but we have several bugs causing either assertion failure or PCC mismatches due to using `ttnn.repeat`. Some examples are [here](https://github.com/tenstorrent/tt-metal/issues/16701) and [here](https://github.com/tenstorrent/tt-metal/issues/16698).

**Comment by patrickroberts** [comment]
Just to clarify, ttnn does not support strided tensor views, so broadcast_to would reallocate unlike torch. Is this just a request for a non-buggy repeat, or a request for an op that creates a strided tensor without reallocating?

**Comment by cmaryanTT** [comment]
@ntarafdar can you comment on repeat status and applicability here

**Comment by ntarafdar** [comment]
repeat on last dim does not work, repeat along with expand will be uplifted for full support by end of month with @jvegaTT 's fix

**Comment by uazizTT** [comment]
> Just to clarify, ttnn does not support strided tensor views, so broadcast_to would reallocate unlike torch. Is this just a request for a non-buggy repeat, or a request for an op that creates a strided tensor without reallocating?

@patrickroberts This is a request for the latter one i.e. an op that creates a strided tensor similar to a tensor view. There are separate issues for non-buggy repeat.

**Comment by eyonland** [comment]
@uazizTT, to avoid any allocations of new buffers and support just a new view of the existing Tensor, our infrastructure would have to provide a definition of a Tensor that includes a stride. 
  We do not carry stride information within our Tensor so we cannot just provide a broadcast view of an existing Tensor.   @ayerofieiev-tt, may have more details as to what is being planned.

**Comment by ayerofieiev-tt** [comment]
So far I think STRIDES can only work for ROW MAJOR tensors.
I currently do not believe that STRIDES work for the general case because of Logical Sharding and Tilization.

CC @sminakov-tt @TT-BrianLiu

**Comment by uazizTT** [comment]
FYI @sdjordjevicTT @AleksKnezevic

**Comment by patrickroberts** [comment]
> So far I think STRIDES can only work for ROW MAJOR tensors. I currently do not believe that STRIDES work for the general case because of
> 
> * Sharding
> * Tilization
> 
> Alignment padding and logical sharding padding make it so many cases require physical data re-arrangement.
> 
> I might be wrong here. Please correct me if you think so
> 
> CC [@sminakov-tt](https://github.com/sminakov-tt) [@TT-BrianLiu](https://github.com/TT-BrianLiu)

Thinking it through, tiled layout is not incompatible with a strided view, if you store the stride information in terms of tiles instead of elements. In BinaryNg, we actually handle broadcasting at the scale of tiles even when broadcasting along H and W, it still works out logically by reusing the entire tile for just one row, column, or element (as long as that row, column, or element gets properly repeated within the entire tile). As for sharding, I'm not sure of the impact on that.

One advantage of strided views in terms of performance, is let's say both tensors of a binary op are broadcasted along C. The output tensor can also be broadcasted along C to represent duplicate computations along that dimension while only having to compute it once for the physical data.

**Comment by eyonland** [comment]
The `TensorLayout::compute_strides(const ttnn::SimpleShape& logical_shape)` is handling the case for the context of alignment and this is a different use-case than what we mean when we talk about strides like we would see in numpy or torch.
To properly handle strides as we understand them in the context of broadcast_to,  we think that the strides themselves have to be stored (not computed) in the Tensor. 

@uazizTT,  my thought is that we move forward with the less than ideal solution for now (ie. creating a new tensor) and circle back to updating this code when strides are properly supported within a Tensor.

**Comment by dchenTT** [comment]
For input and output tensor, I think it make senses for them to have the same memory layout, sharded or interleaved. Is there requirement to support any combination, such as input interleaved output sharded, and input sharded output interleaved?

**Comment by odjuricicTT** [comment]
@dchenTT It would be good to have support for interleaved -> sharded, tho it's not a hard requirement for now.

The more important thing is to clearly assert on all provided memory config-s that are not supported.

**Comment by cmaryanTT** [comment]
@ntarafdar - see above, what's the norm here for other TMs?

**Comment by dchenTT** [comment]
PR is approved, but blocked by post commit test error https://github.com/tenstorrent/tt-metal/issues/20934

**Comment by dchenTT** [comment]
Merged. https://github.com/tenstorrent/tt-metal/pull/19759

**Comment by dchenTT** [comment]
Basic support for the op was merged long time ago.
